File: static/python/morris_traversal.py
from binary_tree_prototype import BinaryTreeNode

result = []


# @include
def inorder_traversal(tree):
    it = tree
    while it:
        pre = it.left
        if pre:
            # Finds the predecessor of it.
            while pre.right and pre.right is not it:
                pre = it.left = pre.right

            # Processes the successor link.
            if pre.right:  # pre.right is it.
                # Reverts the successor link if predecessor's successor is it.
                pre.right = None
                print(it.data)
                # @exclude
                result.append(it.data)
                # @include
                it = it.right
            else:  # If predecessor's successor is not it.
                pre.right = it
                it = it.left

        else:
            print(it.data)
            # @exclude
            result.append(it.data)
            # @include
            it = it.right
# @exclude


def main():
    #      3
    #    2   5
    #  1    4 6
    tree = BinaryTreeNode(3)
    tree.left = BinaryTreeNode(2)
    tree.left.left = BinaryTreeNode(1)
    tree.right = BinaryTreeNode(5)
    tree.right.left = BinaryTreeNode(4)
    tree.right.right = BinaryTreeNode(6)
    # should output 1 2 3 4 5 6
    inorder_traversal(tree)
    golden_res = [1, 2, 3, 4, 5, 6]
    assert result == golden_res


if __name__ == '__main__':
    main()

File: static/cpp/Morris_traversal.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <iostream>
#include <memory>
#include <vector>

#include "./Binary_tree_prototype.h"

using std::cout;
using std::endl;
using std::make_unique;
using std::unique_ptr;
using std::vector;

vector<int> result;

// @include
void InorderTraversal(const unique_ptr<BinaryTreeNode<int>>& tree) {
  auto* iter = tree.get();
  while (iter) {
    if (iter->left.get()) {
      // Finds the predecessor of iter.
      auto& pre = iter->left;
      while (pre->right && pre->right.get() != iter) {
        pre.reset(pre->right.get());
      }

      // Processes the successor link.
      if (pre->right) {  // pre->right.get() == iter.
        // Reverts the successor link if predecessor's successor is iter.
        pre->right.release();
        cout << iter->data << endl;
        // @exclude
        result.emplace_back(iter->data);
        // @include
        iter = iter->right.get();
      } else {  // If predecessor's successor is not iter.
        pre->right.reset(iter);
        iter = iter->left.get();
      }
    } else {
      cout << iter->data << endl;
      // @exclude
      result.emplace_back(iter->data);
      // @include
      iter = iter->right.get();
    }
  }
}
// @exclude

int main(int argc, char* argv[]) {
  //      3
  //    2   5
  //  1    4 6
  unique_ptr<BinaryTreeNode<int>> tree =
      make_unique<BinaryTreeNode<int>>(BinaryTreeNode<int>{3});
  tree->left = make_unique<BinaryTreeNode<int>>(BinaryTreeNode<int>{2});
  tree->left->left = make_unique<BinaryTreeNode<int>>(BinaryTreeNode<int>{1});
  tree->right = make_unique<BinaryTreeNode<int>>(BinaryTreeNode<int>{5});
  tree->right->left =
      make_unique<BinaryTreeNode<int>>(BinaryTreeNode<int>{4});
  tree->right->right =
      make_unique<BinaryTreeNode<int>>(BinaryTreeNode<int>{6});
  // should output 1 2 3 4 5 6
  InorderTraversal(tree);
  vector<int> golden_res = {1, 2, 3, 4, 5, 6};
  assert(equal(result.begin(), result.end(), golden_res.begin(),
               golden_res.end()));
  return 0;
}
